ovs-update.md

```diff
@@ -0,0 +1,281 @@
+pods to be started. This will necessarily restart the `ovs-vswitchd` processes.
+
+Right now, there is no provision to mitigate the effect of this restart.
+Meaning, the new process will start with empty OpenFlow tables and will
+invalidate datapath flows. It will wait for `ovn-controller` to reconnect and
+reconfigure it via OpenFlow control socket. In the meantime, the traffic may be
+disrupted.
+
+### Proposal
+
+To mitigate the effect of `ovs-vswitchd` restart, we can utilize
+`other_config:flow-restore-wait` mechanism in Open vSwitch.
+[This mechanism](https://github.com/openvswitch/ovs/blob/fbade819d2de1685c49e1deaf62d45049a5c2a27/vswitchd/vswitch.xml#L104)
+allows Open vSwitch to restore the OpenFlow tables from a backup file after a
+restart.
+
+This mechanism is already utilized in the Open vSwitch RPM package when doing
+`systemctl reload openvswitch`. Since systemd unit files are not used to manage
+`ovs-vswitchd` in pods, we should implement this mechanism in shutdown and
+startup scripts used for OVS DaemonSet.[^0]
+
+The proposed implementation would work as follows:
+
+- Before shutting down `ovs-vswitchd`,
+  [`PreStop` hook](https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/)
+  for OVS DaemonSet pods would set `other_config:flow-restore-wait` to `true`
+  in the OVSDB. This will indicate to the new `ovs-vswitchd` process that it
+  should not flush the kernel datapath or connect to its controllers until
+  saved flows are restored.
+
+- In addition, the `PreStop` hook would save the OpenFlow tables to a file in
+  the pod's volume. This can be done with `ovs-ofctl dump-flows` command.
+
+- After the new `ovs-vswitchd` process starts in the new pod, the startup script will check
+  for the presence of the saved flows file. If it exists, it will restore the
+  flows using `ovs-ofctl` command. It will then remove the file.
+
+- Once the flows are restored, the startup script will set
+  `other_config:flow-restore-wait` back to `false` in the OVSDB. This will
+  allow the new `ovs-vswitchd` process to connect to its controllers and start
+  processing packets.
+
+## Bare metal Open vSwitch for data plane nodes
+
+In contrast to podified Open vSwitch for OVN gateways, Open vSwitch on data
+plane nodes is deployed on bare metal using RPM packages. The deployment is
+managed by `edpm_ovs` role.
+
+### RPM packaging
+
+There are several important details to consider when updating Open vSwitch RPM
+packages:
+
+- In addition to the `openvswitch*` package, OpenStack also ships a
+  `<distro>-openvswitch` wrapper package. (`rdo-openvswitch` for RDO,
+  `rhosp-openvswitch` for RHOSO). This package is used to select the Open
+  vSwitch "stream" to use (e.g. `3.0` vs `3.3`). It is needed because all Open
+  vSwitch packages are shipped in the same repository.[^1] Without the wrapper,
+  we would always install the latest version from the repository, which is not
```

Comment:
I will remove the sentence.

So what's then the reason to have the wrapper? To untangle neutron dependency on openvswitch from the exact stream choice?

Reply:
it too assists in upgrade cases to new streams

Reply:
@karelyatin but how exactly does it assist? couldn't neutron package directly pick the needed openvswitch stream via its dependency? I'm legit trying to understand the special role of the package that justifies its existance.

Reply:
@booxter yes it's possible to install the required streams directly with dependencies, and for containers it would work as we rollout containers(and container builds will install(no update/upgrade) the required stream as per dependencies). But consider the non-container cases it would not work, the stream in itself conflict with each other i.e ovs3.2 conflicts with ovs3.3 and can't be upgraded(can try install and update to other). The wrapper package allows such cases

Reply:
@booxter Yes i think it would fail as it will try to check/install ovs3.3 and would conflict with current installed ovs3.1. I am unsure about the orphan dependency, but i believe it cleans up, even if that happens it would happen later but it would even before.  To be sure you can try that if you want, would be interested if that works somehow.

Reply:
OK I understand the concern now. I will eventually try it (would need to build some custom neutron packages with explicit openvswitch* dependencies to validate it.)